RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,823 @@
+```
+
+This will not generate a warning, but it is unsafe and can be the cause of a `NullReferenceException` if the reference was indeed `null` under some condition.
+
+#### Warnings
+
+In all other situations not covered by flow analysis, a warning is emitted by the compiler if a nullable reference is dereferenced or casted to a non-null type:
+
+```fsharp
+let f (ns: string | null) =
+    printfn "%d" ns.Length // WARNING: 'ns' may be null
+
+    let mutable ns2 = ns
+    printfn "%d" ns2.Length // WARNING: 'ns2' may be null
+
+    let s = ns :> string // WARNING: 'ns' may be null
+```
+
+A warning is given when casting from `'S[]` to `('T | null)[]` and from `('S | null)[]` to `'T[]`.
+
+A warning is given when casting from `C<'S>` to `C<'T | null>` and from `C<'S | null>` to `C<'T>`.
+
+### Checking of non-null references
+
+#### Null assignment and passing
+
+A warning is given if `null` is assigned to a non-null value or passed as a parameter where a non-null reference type is expected:
+
+```fsharp
+let mutable s: string = "hello"
+s <- null // WARNING
+
+let s2: string = null // WARNING
+
+let f (s: string) = ()
+f null // WARNING
+```
+
+#### F# Collection initialization
+
+A warning is given if `null` is used in an F# array/list/seq expression, where the type is already annotated to be non-nullable:
+
+```fsharp
+let xs: string[] = [| ""; ""; null |] // WARNING
+let ys: string list = [ ""; ""; null ] // WARNING
+let zs: seq<string> = seq { yield ""; yield ""; yield null } // WARNING
+```
+
+#### Constructor initialization
+
+Today, it is already a compile error if all fields in an F# class are not initialized by constructor calls. This behavior is unchanged. See [Compatibility with Microsoft.FSharp.Core.DefaultValueAttribute](FS-1060-nullable-reference-types.md#compatibility-with-microsoft.fsharp.core.defaultValueAttribute) for considerations about code that uses `[<DefaultValue>]` with reference types today, which produces `null`.
+
+### Generics
+
+A type parameter is assumed to have non-nullable constraints when declared "normally":
+
+```fsharp
+type C<'T>() = // 'T is non-nullable
+class end
+```
+
+A warning is given if a nullable reference type is used to parameterize another type `C` that  does not accept nullable reference types:
+
+```fsharp
+type C<'T>() = class end
+
+let c = C<string | null> // WARNING
+```
+
+#### Constraints
+
+Today, there are two relevant constraints in F# - `null` and `not struct`:
+
+```fsharp
+type C1<'T when 'T: null>() = class end
+type C2<'T when 'T: not struct>() = class end
+```
+
+Both are actually the same thing for interoperation purposes, as they emit `class` as a constraint. See [Unresolved questions](nullable-reference-types.md#unresolved-questions) for more.
+
+If an F# class `B` has nullability as a constraint, then an inheriting class `D` also inherits this constraint, as per existing inheritance rules:
+
+```fsharp
+type B<'T | null>() = class end
```

Comment:
Per comment above, I don't know what this `type B<'T | null>` means, or at least don't expect to have it in the language.  If, per above, allowing nullable instantiations of generic things is the default, then it's implied.